ls1intum/Artemis
# [Feature] Grading with Static Code Analysis

### Checklist

### Motivation and Context



This PR integrates static code analysis into the grading system.

### Description


- Server
    - Score Generation: Applying penalties from SCA categories for SCA feedbacks
    - Filter feedback for invisible SCA categories
    - The 'updatedTestCases' flag of a programming exercise is now also used to indicate updated categories
- Client
    - SCA categories are loaded from the server
    - The 'test-case-service' is now named 'grading-service'
    - Save & Reset Buttons are now separate for both tables (test-cases and categories)
    - All grading-related messages are adjusted to include the category variant.
    - The category table is updated to handle the category state
    - The 'updatedTestCase' flag is inside the component used as 'updatedGradingConfig'

### Steps for Testing


1. Log in to Artemis
2. Create a Programming Exercise with Static Code Analysis enabled
3. Configure Test Cases and SCA Categories through the grading page.
4. Participate in the exercise and view the result + feedback.
5. Edit the grading configuration, recalculate the score and view the result change as a student.
    - Set more categories to 'Graded', apply higher penalties and re-evaluate the scores -> Score should decrease
    - Set categories to 'Inactive' -> Feedbacks in this category should not be visible anymore
    - Set categories back to 'Feedback' or 'Graded' and re-trigger all builds -> Feedbacks should be visible again

### Test Coverage


<img width="1431" alt="Bildschirmfoto 2020-10-03 um 15 55 46" src="https://user-images.githubusercontent.com/13920539/94996142-1e3c2b80-05a3-11eb-827f-35e1d3566b59.png">
<img width="1189" alt="Bildschirmfoto 2020-10-03 um 18 00 31" src="https://user-images.githubusercontent.com/13920539/94996144-1f6d5880-05a3-11eb-9cfe-757f41d6667c.png">
<img width="1185" alt="Bildschirmfoto 2020-10-03 um 18 00 38" src="https://user-images.githubusercontent.com/13920539/94996145-2005ef00-05a3-11eb-9c0b-bf9d67a45ae0.png">

### Screenshots



<img width="1413" alt="Bildschirmfoto 2020-10-05 um 22 34 07" src="https://user-images.githubusercontent.com/13920539/95129169-e972d480-075a-11eb-909e-12acc0056304.png">
<img width="1672" alt="95126332-95fe8780-0756-11eb-8707-00186240af03" src="https://user-images.githubusercontent.com/13920539/95228854-1dec9c00-0800-11eb-9a46-7061abfeaecb.png">
